### PR TITLE
Add basic NetworkPolicy

### PR DESCRIPTION
**Pull request by hibell**
Network policies only added if createNetworkPolicies is enabled.

**What this PR does / why we need it?**:

- Adds network policies.

**Does this PR introduce a user-facing change?**

- [ ] User guide
- [ ] `CHANGELOG.md`

**Which issue(s) this PR fixes**:
N/A

### CODE REVIEW

File: api/v1beta2/runtimecomponent_types.go [hunk 1]
```diff
@@ -357,7 +366,7 @@ const (
 // +kubebuilder:printcolumn:name="Reason",type="string",JSONPath=".status.conditions[?(@.type=='Reconciled')].reason",priority=1,description="Reason for the failure of reconcile condition"
 // +kubebuilder:printcolumn:name="Message",type="string",JSONPath=".status.conditions[?(@.type=='Reconciled')].message",priority=1,description="Failure message from reconcile condition"
 // +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",priority=0,description="Age of the resource"
-//+operator-sdk:csv:customresourcedefinitions:displayName="RuntimeComponent",resources={{Deployment,v1},{Service,v1},{StatefulSet,v1},{Route,v1},{HorizontalPodAutoscaler,v1},{ServiceAccount,v1},{Secret,v1}}
+// +operator-sdk:csv:customresourcedefinitions:displayName="RuntimeComponent",resources={{Deployment,v1},{Service,v1},{StatefulSet,v1},{Route,v1},{HorizontalPodAutoscaler,v1},{ServiceAccount,v1},{Secret,v1}}
```
Comment (leochr) [inline api/v1beta2/runtimecomponent_types.go:369]: Let's add `{NetworkPolicy,v1}` so it's shown under the 'Resources' tab of the custom resource (in OpenShift UI)

File: api/v1beta2/runtimecomponent_types.go [hunk 2]
```diff
@@ -232,6 +235,12 @@ type RuntimeComponentService struct {
 	Bindable *bool `json:"bindable,omitempty"`
 }
 
+// RuntimeComponentNetworkPolicy configures the network policy
```
Comment (leochr) [inline api/v1beta2/runtimecomponent_types.go:238]: These descriptions end up in the CRD and are also used in the OpenShift UI.
Suggestion: Defines the network policy

The child fields also require descriptions

**Review by leochr — COMMENTED** [review]
@hibell Thank you. Added some comments. Only managed to review half of it. I'll review the rest by tomorrow.

File: bundle/tests/scorecard/kuttl/network-policy/01-assert.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+timeout: 60
+---
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: network-policy-rc
+spec:
+  ingress:
+    - from:
+        - namespaceSelector: {}
```
Comment (leochr) [inline bundle/tests/scorecard/kuttl/network-policy/01-assert.yaml:12]: Is it necessary for the port(s) to be specified for Route/Ingress to work? If not, we should only specify ports when fromLabels is specified (to allow traffic from certain pods).
Also, is traffic restricted for all namespaces, including the CR namespace?

File: bundle/tests/scorecard/kuttl/network-policy/03-assert.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: network-policy-rc
+spec:
+  ingress:
+    - from:
+        - namespaceSelector: {}
+      ports:
+        - port: 8080
```
Comment (leochr) [inline bundle/tests/scorecard/kuttl/network-policy/03-assert.yaml:14]: Kuttl wouldn't assert that 9090 and 9091 are no longer specified. We need an 'error' check on those ports to make sure they no longer exist.

**Review by leochr — COMMENTED** [review]
@hibell Thanks for the updates. Added some comments to clarify some behaviour.

File: bundle/tests/scorecard/kuttl/network-policy/01-assert.yaml (hunk 3 above)
Comment (hibell) [reply bundle/tests/scorecard/kuttl/network-policy/01-assert.yaml:12]: > Also, is traffic restricted for all namespaces, including the CR namespace?

Yes. With namespaceSelector: {} it should apply to all ingress traffic from any namespace.

> Is it necessary for the port(s) to be specified for Route/Ingress to work? If not, we should only specify ports when fromLabels is specified (to allow traffic from certain pods).

Yes, without the ports specified, external traffic would be blocked.

**Comment by hibell** [comment]
@leochr Made the changes requested during yesterday's code review.

With the following RCO CR:

```yaml
---
apiVersion: rc.app.stacks/v1beta2
kind: RuntimeComponent
metadata:
  name: inventory
spec:
  applicationName: monitor-app
  applicationImage: hwibell/inventory-service:latest
  expose: true
  service:
    port: 9080
    ports:
      - port: 9443
        protocol: TCP
---
apiVersion: rc.app.stacks/v1beta2
kind: RuntimeComponent
metadata:
  name: system
spec:
  applicationName: monitor-app
  applicationImage: hwibell/system-service:latest
  service:
    port: 9080
    ports:
      - port: 9443
        protocol: TCP
```

we get the following NetworkPolicies:

```yaml
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: inventory
  namespace: guide
spec:
  ingress:
  - from:
    - namespaceSelector: {}
    ports:
    - port: 9080
      protocol: TCP
    - port: 9443
      protocol: TCP
  podSelector:
    matchLabels:
      rc.app.stacks/name: inventory
  policyTypes:
  - Ingress
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: system
  namespace: guide
spec:
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: guide
      podSelector:
        matchLabels:
          app.kubernetes.io/part-of: monitor-app
    ports:
    - port: 9080
      protocol: TCP
    - port: 9443
      protocol: TCP
  podSelector:
    matchLabels:
      rc.app.stacks/name: system
  policyTypes:
  - Ingress
```

* With the above network policies, the exposed `inventory` app accepts external requests and it is able to communicate with the `systems` service via `system.guide.svc`:

```console
$ curl -sD - http://inventory-guide.apps.hyperion-cnb.cp.fyre.ibm.com/inventory/systems/system.guide.svc
HTTP/1.1 200 OK
content-type: application/json
content-language: en-US
content-length: 4171
date: Fri, 08 Apr 2022 04:51:07 GMT
set-cookie: 73540f595aee9b387fce99dd27ed7ceb=433e7166141511bb9233fb76f130a990; path=/; HttpOnly
cache-control: private
...
```

* Attempting to hit to the `system` service from a different pod in the namespace results in ingress traffic being denied:

```console
$ oc exec ferret-sample-688f8cdb58-fgs27 -c ferret-sample  -- curl --max-time 5 -sD - http://system.guide.svc:9080 -o /dev/null
command terminated with exit code 28
```

* Attempting to hit to the `system` service from the inventory pod in the namespace results in ingress traffic being allowed:

```console
$ oc exec $(oc get pods -l app.kubernetes.io/name=inventory -o=jsonpath='{.items[].metadata.name}') -- curl --max-time 5 -sD - http://system.guide.svc:9080 -o /dev/null
HTTP/1.1 200 OK
Date: Fri, 08 Apr 2022 04:55:38 GMT
Content-Type: text/html
Last-Modified: Thu, 07 Apr 2022 19:08:16 GMT
Content-Length: 2487
Content-Language: en-US
```

**Review by leochr — APPROVED** [review]
@hibell Looks good! Thanks

**Pull request merged**